### ISSUE TITLE
Random segfault after fio_flush_all

### ISSUE THREAD

**Issue by ianks** [boazsegev/iodine#69]
Hey @boazsegev, hope all is well :smile: 	

We are getting a random segfault after an acceptance tests finish in capybara. I believe capybara may be killing the Iodine thread (which is should do automatically), at which point Iodine sometimes segfaults in `fio_flush_all`

Here is the backtrace info:
```
ruby 2.6.2p47 (2019-03-13 revision 67232) [x86_64-linux]

-- Machine register context ------------------------------------------------
 RIP: 0x00007f98fb43063e RBP: 0x00007f98eb9fd6d0 RSP: 0x00007f98eb9fd6b0
 RAX: 0x0000000000000000 RBX: 0x0000000000000000 RCX: 0x00007f993bc10010
 RDX: 0xf2f0ab6d387d4400 RDI: 0x0000000000000000 RSI: 0xf2f0ab6d387d4400
  R8: 0x00007f98eb9ff700  R9: 0x00007f98eb9ff700 R10: 0xfffffffffffff942
 R11: 0x0000000000000000 R12: 0x00007f98ec5f9ace R13: 0x00007f98ec5f9acf
 R14: 0x00007f98eb9ff700 R15: 0x0000000000000000 EFL: 0x0000000000010202

-- C level backtrace information -------------------------------------------

bin/paraspec [worker-6](0x262d84) [0x55b29fc4bd84]
bin/paraspec [worker-6](0x259090) [0x55b29fc42090]
bin/paraspec [worker-6](0x138612) [0x55b29fb21612]
/usr/lib/libpthread.so.0(__restore_rt+0x0) [0x7f99036f33c0]
/home/ianks/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/iodine-0.7.27/lib/iodine/iodine.so(fio_flush_all+0x3e) [0x7f98fb43063e]
/home/ianks/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/iodine-0.7.27/lib/iodine/iodine.so(0xa2bef) [0x7f98fb4b9bef]
/usr/lib/libpthread.so.0(0x7a9d) [0x7f99036e8a9d]
/usr/lib/libc.so.6(clone+0x43) [0x7f9902f21b23]
```

I'm more than happy to provide other infomation that might be helpful as well, but my C is a little rusty :smile:

**Comment by boazsegev** [comment]
#### TL;DR;

I **suspect** that this might not be an iodine bug as much as it might be a result of an ungraceful (mishandled) iodine shutdown.

Please correct me if you think I might be mistaken.

I'll try to push a fix that will prevent this error even when shutdown is mishandled, but a mishandled shutdown should be avoided (this isn't the only possible error).

#### The long version:

It's never a good idea to kill threads unexpectedly, especially if they aren't exclusively Ruby threads:

- Although Ruby **might** protect it's threads and data structures from being killed within Ruby's critical sections, the OS does NOT.

- Unlike exclusively Ruby threads, iodine threads manage locks and other critical section data while outside of Ruby's GIL.

- Hence, killing the iodine thread unexpectedly might produce undefined behavior. For example, the memory allocation system (`malloc`) might be corrupted by unreleased locks or incomplete data structure updates.

This error is probably a good example of such an occurrence, albeit an iodine specific occurrence:

- Iodine spawns an IO thread that runs periodically outside the Ruby GIL (global interpreter lock). This thread flushes out any pending outgoing IO, protecting against user code that might block all of the application's threads.

- Normally, the IO thread is signaled during iodine's shutdown process and killed gracefully. However, when the main iodine thread is killed ungracefully, the IO thread might keep running in the background.

- Since the IO thread doesn't exit, it's still alive when the library cleanup occurs during the process's final cleanup stage... resulting in the IO thread attempting to access memory that was freed during the cleanup process.

**If I am correct**, this would only happen when the IO thread's periodical "flushing" occurs right after iodine's final cleanup but before the process finishes other cleanup tasks and exists...?

**Comment by boazsegev** [comment]
P.S.

You can try installing from Github's main branch to see if the patch I pushed solves this. If it does, then my suspicions about the source of the issue were correct.

**Comment by boazsegev** [comment]
Any updates?

Please let me know if I was wrong about the source of the issue - since I can't reproduce the issue, I'm not sure.

**Comment by ianks** [comment]
yes let me try :)

**Comment by boazsegev** [comment]
Any updates on this? Does the updated version (0.7.33) solve the issue?
